include/lsst/afw/fitsCompression.h

```diff
@@ -0,0 +1,488 @@
+      case -64: return std::make_shared<PixelArray<boost::float64_t>>(array);
+      default:
+        std::ostringstream os;
+        os << "Unrecognized bitpix: " << bitpix;
+        throw LSST_EXCEPT(pex::exceptions::InvalidParameterError, os.str());
+    }
+}
+
+} // namespace detail
+
+
+/// Options for tile compression of image pixels
+///
+/// Tile compression is a feature provided by cfitsio, where contiguous parts
+/// of the image ("tiles", e.g., rows, multiple rows, blocks or the entire image)
+/// are compressed separately. The aim of this struct is to hold the parameters
+/// used to configure the compression.
+///
+/// Floating-point images can be losslessly compressed (quantizeLevel=0.0) using
+/// (only) the GZIP or GZIP_SHUFFLE compression schemes, but the compression
+/// factor so achieved is modest (e.g., ~ 10% compression). Better compression
+/// factors can be achieved if floating-point images are first quantised into
+/// integer images. This can be done by cfitsio (through the quantizeLevel parameter)
+/// or through use of the ImageScalingOptions.
+///
+/// The Compression is specified by:
+/// * the compression scheme (what algorithm is used to do the actual compression)
+/// * the tile size
+/// * the quantization level (for quantization applied by cfitsio; for floating-point images)
+///
+/// Due to bugs, cfitsio may require setting the quantizeLevel to a value other than
+/// zero when compressing integer data, but in this case it should have no effect.
+struct ImageCompressionOptions {
+    /// Compression algorithms
+    ///
+    /// cfitsio's compression schemes are \#defines; these have a namespace.
+    ///
+    /// We deliberately don't support HCOMPRESS: it doesn't appear to be useful to us (e.g., lossy)
+    /// and it requires extra configuration.
+    enum CompressionScheme {
+        NONE,              ///< No compression
+        GZIP,              ///< Standard GZIP compression
+        GZIP_SHUFFLE,      ///< GZIP compression with shuffle (most-significant byte first)
+        RICE,              ///< RICE compression
+        PLIO,              ///< PLIO compression
+    };
+    typedef ndarray::Array<long, 1> Tiles;
+
+    CompressionScheme scheme;  ///< Compresion algorithm to use
+    Tiles tiles;           ///< Tile size; a dimension with 0 means infinite (e.g., to specify one row: 0,1)
+    float quantizeLevel;   ///< quantization level: 0.0 = none requires use of GZIP or GZIP_SHUFFLE
+
+    /// Custom compression
+    explicit ImageCompressionOptions(
+        CompressionScheme scheme_,
+        Tiles tiles_,
+        int quantizeLevel_=0.0
+    ) : scheme(scheme_), tiles(tiles_), quantizeLevel(quantizeLevel_) {}
+
+    /// Compression by rows or entire image
+    ///
+    /// @param[in] scheme_  Compression algorithm to use
+    /// @param[in] rows  Number of rows per tile (0 = entire image)
+    /// @param[in] quantizeLevel_  cfitsio quantization level
```

Comment:
In FITS, rows are contiguous (indexing `[y*nx + x]`). Our coadds are done by rows (well, that's now the default; it's possible to do them in blocks, but we've set the default to work by rows because FITS works by rows). Read patterns for other operations (e.g., creating cutouts) will differ. We can do blocks, but I think rows is a decent default, it's easy to specify, and it's what Pan-STARRS uses. There is another constructor to allow blocks to be specified if the user desires.